oauthproxy.go

```diff
@@ -838,6 +839,27 @@ func (p *OAuthProxy) enrichSessionState(ctx context.Context, s *sessionsapi.Sess
 func (p *OAuthProxy) AuthOnly(rw http.ResponseWriter, req *http.Request) {
 	session, err := p.getAuthenticatedSession(rw, req)
 	if err != nil {
+		if authOnlyRedirectSignIn(req) {
+			if isAjax(req) {
+				// no point redirecting an AJAX request
+				p.errorJSON(rw, http.StatusUnauthorized)
+				return
+			}
+
+			// Orignal request URI is captured and added using `rd` parameter in the redirect
+			// to ensure a redirect to the original URL will still happen after OAuth/Signin process
+			redirect, err := p.getAppRedirect(req)
```

Comment:
`getAppRedirect` has a lot of logic paths to figure out where the app redirect goes to -- In traefik will this come from the `X-Forwarded-(Proto|Host|Uri)` headers?

If so and we then map them to `rd` on the SignIn handler -- that is acceptable (or is traffic still sending the same X headers to the sign in..?)

FYI - `rd` takes priority over everything else -- so this logic seems good to me. I just wanted to talk it out in case there's unintentional mixing of logic.

Reply:
Yes, Traefik uses the `X-Forwarded-(Proto|Host|Uri)` headers to provide the request URL information. The `getAppRedirect` seems to handle all the various cases nicely and the usage of `rd` parameter seemed like the best case here to preserve the original request URL

The flow is:
- original unauthenticated request comes in to `https://a-service.example.com/`
- Traefik ForwardAuth middleware forwards this request to `https://oauth2-proxy.example.com/oauth2/auth?redirect_signin=true`
- As per new logic, the original request URL is captured in `rd=https://a-service.example.com/` and added as a QS to `/oauth2/sign_in` redirect path with a 302 redirect returned to the client